### PR TITLE
[Rebase & FF][202405] SecurityLockLib to log security milestones. 

### PR DESCRIPTION
**Pull request by Flickdm**
This PR  Adds the ability to record security lock events (software vs hardware) Additionally, logs when capsule interface is locked

### Overview of commits:

[Implemented SecurityLockLib to log security milestones.](https://github.com/microsoft/mu_basecore/pull/1032/commits/8a8f7824caf0084012598a04bba736809a7fe089)
[MdeModulePkg: Add Locking support for Capsule Interface](https://github.com/microsoft/mu_basecore/pull/1032/commits/d781d56005c4322be882b19dc326b8d03636bd19)

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

----
### How This Was Tested
From 2311.

### CODE REVIEW

File: MdeModulePkg/Core/PiSmmCore/PiSmmIpl.c [hunk 1]
```diff
@@ -34,9 +34,12 @@
 #include <Library/UefiRuntimeLib.h>
 #include <Library/PcdLib.h>
 #include <Library/ReportStatusCodeLib.h>
-#include "PiSmmCorePrivateData.h"
+
+#include <Library/SecurityLockAuditLib.h> // MU_CHANGE
 #include <Library/SafeIntLib.h>
 
+#include "PiSmmCorePrivateData.h"
```
Comment (os-d) [inline MdeModulePkg/Core/PiSmmCore/PiSmmIpl.c]: MU_CHANGE here

File: MdeModulePkg/Include/Library/SecurityLockAuditLib.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+This library implements the necessary functions
+to log hardware and software security locks for post-processing
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __SECURITY_LOCK_LIB_H__
```
Comment (os-d) [inline MdeModulePkg/Include/Library/SecurityLockAuditLib.h]: nit: `SECURITY_LOCK_LIB_H_`

File: MdeModulePkg/Include/Library/SecurityLockAuditLib.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
+/**
+
+This library implements the necessary functions
+to log hardware and software security locks for post-processing
+
+Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (os-d) [inline MdeModulePkg/Include/Library/SecurityLockAuditLib.h]: Copyright (c) Microsoft Corporation.

Please do for other files as well.

File: MdeModulePkg/Include/Library/SecurityLockAuditLib.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef __SECURITY_LOCK_LIB_H__
+#define __SECURITY_LOCK_LIB_H__
+
+#define SECURITY_LOCK_REPORT_EVENT(LockMessage, LockType) \
+      SecurityLockReportEvent(&gEfiCallerIdGuid,__FUNCTION__,LockMessage,LockType); \
```
Comment (os-d) [inline MdeModulePkg/Include/Library/SecurityLockAuditLib.h]: `__func__`

File: MdeModulePkg/Library/SecurityLockAuditDebugMessageLib/SecurityLockAuditDebugMessageLib.c [hunk 5]
```diff
@@ -0,0 +1,47 @@
+/** @file
+  This library implements the necessary functions
+  to log hardware and software security locks for post-processing
+
+  Copyright (c) Microsoft Corporation. All rights reserved.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Base.h> // MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/Library/SecurityLockAuditDebugMessageLib/SecurityLockAuditDebugMessageLib.c]: nit: whole file is mu change, don't need this

File: MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.inf [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+[Sources.common]
+  SecurityLockAuditLibNull.c
+
+[Packages]
+  MdePkg/MdePkg.dec
+  MdeModulePkg/MdeModulePkg.dec
+
+[LibraryClasses]
+  BaseLib
```
Comment (os-d) [inline MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.inf]: add empty line at EOF

File: MdeModulePkg/MdeModulePkg.dec [hunk 7]
```diff
@@ -163,6 +163,9 @@
   ##  @libraryclass  Provides services to access UEFI variable flash information.
   #
   VariableFlashInfoLib|Include/Library/VariableFlashInfoLib.h
+  ##  @libraryclass  Provides a way for logging security locks
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dec:167]: nit: add an empty line separating from the last entry

File: MdeModulePkg/MdeModulePkg.dsc [hunk 8]
```diff
@@ -232,8 +233,10 @@
   MdeModulePkg/Application/HelloWorld/HelloWorld.inf
   MdeModulePkg/Application/DumpDynPcd/DumpDynPcd.inf
   MdeModulePkg/Application/MemoryProfileInfo/MemoryProfileInfo.inf
-  MdeModulePkg/Library/ParallelLzmaCustomDecompressLib/ParallelLzmaCustomDecompressLib.inf       ## MU_CHANGE
   MdeModulePkg/Library/MemoryTypeInformationChangeLibNull/MemoryTypeInformationChangeLibNull.inf ## MU_CHANGE
+  MdeModulePkg/Library/SecurityLockAuditDebugMessageLib/SecurityLockAuditDebugMessageLib.inf     ## MU_CHANGE
+  MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.inf                     ## MU_CHANGE
+  MdeModulePkg/Library/ParallelLzmaCustomDecompressLib/ParallelLzmaCustomDecompressLib.inf       ## MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dsc]: Automerge moved the decompresslib, let's leave that where it was, so git blame is accurate

File: MdeModulePkg/MdeModulePkg.dsc [hunk 9]
```diff
@@ -564,5 +567,4 @@
 [Components.X64]
   MdeModulePkg/Universal/CapsulePei/CapsuleX64.inf
 
-[BuildOptions]
-
+[BuildOptions]
```
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dsc]: Drop this

**Review by os-d — APPROVED** [review]

**Review by makubacki — APPROVED** [review]

File: MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.c [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+VOID
+EFIAPI
+SecurityLockReportEvent (
+  IN GUID         *Module,
+  IN CONST CHAR8  *Function,
+  IN CONST CHAR8  *LockEventText,
+  IN LOCK_TYPE    LockType
+  )
+{
+}
```
Comment (makubacki) [inline MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.c]: Our normal convention would not leave the function body empty.

```suggestion
{
  return;
}
```

File: MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.inf [hunk 11]
```diff
@@ -0,0 +1,26 @@
+##
+#
+#  Null library for security lock logging that does nothing but meet compile requirements
+#
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+##
```
Comment (makubacki) [inline MdeModulePkg/Library/SecurityLockAuditLibNull/SecurityLockAuditLibNull.inf]: Nit: Can you clean up some of the extra space in these file headers? This PR is a good time for tidying up like this.

**Pull request merged**